Sprint 3/step 1

Ekaterina64 opened the pull request.

Ekaterina-ux committed:
Форматирование верстки

Ekaterina-ux committed:
Улучшено:
- Очистка конструктора по закрытию модального окна
- Использование в useSelector именованной функции
- Мемоизация

Ekaterina-ux committed:
Страницы login, register, forgot/reset password,
404

Ekaterina-ux committed:
Регистрация + запрос на восстановление пароля

Ekaterina-ux committed:
Доработки по роутингу

andryxxa93 commented on src/components/app/app.jsx:
```diff
@@ -1,24 +1,125 @@
+
+	const forgotPasswordSuccess = useSelector(getForgotPasswordSuccess)
+	const user = useSelector(getUser)
+	const access = getCookie('accessToken')
```
**Можно лучше**
Так как строки accessToken и refreshToken часто в повторяются в коде их стоит вынести в отдельные константы

andryxxa93 commented on src/pages/forgot-password.jsx:
```diff
@@ -0,0 +1,62 @@
+	const dispatch = useDispatch()
+	const forgotPasswordSuccess = useSelector(getForgotPasswordSuccess)
+
+	const handleChange = e => {
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:
```js
 export function useForm(inputValues) {
   const [values, setValues] = useState(inputValues);
 
   const handleChange = (event) => {
     const {value, name} = event.target;
     setValues({...values, [name]: value});
   };
   return {values, handleChange, setValues};
 }
```
    
Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

 И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:
    
 ```js
   const {values, handleChange, setValues} = useForm({});
 ```

andryxxa93 commented on src/pages/profile/profile.jsx:
```diff
@@ -0,0 +1,152 @@
+	}
+	return (
+		<section className={styles.main}>
+			<nav className={`${styles.sidebar} mr-15`}>
```
**Можно лучше**
Компонент лучше декомпозиировать и сделать из навигации внутри страницы профиля отдельный компонент

andryxxa93 reviewed (commented):
Здравствуйте! Проделана большая работа 👍 

Сильные стороны работы:
- Хорошая работа с формами
- Правильно настроен роутинг модальных окон
- Роуты записаны в константы
- Настроены все редиректы
- Есть автологин

Однако есть пару моментов, которые необходимо исправить, чтобы работа была принята.

Необходимо исправить:
- При запросах, использующих `accessToken`, должен проверятся ответ сервера. Если время жизни токена истекло, то необходимо выполнить запрос обновления токена и повторить запрос, который не был выполнен успешно. Воспроизвести проблему можно авторизовавшись и перейдя в личный кабинет подождать 20 минут и попробовать изменить данные пользователя.

Для этого используйте обертку над fetch, примерно такую

```
const checkReponse = (res) => {
  return res.ok ? res.json() : res.json().then((err) => Promise.reject(err));
};

const fetchWithRefresh = async (url, options) => {
  try {
    const res = await fetch(url, options); //делаем запрос
    return await checkReponse(res);
  } catch (err) {
    if (err.message === "jwt expired") {
      const refreshData = await refreshToken(); //обновляем токен
      localStorage.setItem("refreshToken", refreshData.refreshToken);
      localStorage.setItem("accessToken", refreshData.accessToken); //(или в cookies)
      options.headers.authorization = refreshData.accessToken;
      const res = await fetch(url, options); //вызываем перезапрос данных
      return await checkReponse(res);
    } else {
      return Promise.reject(err);
    }
  }
};
```

Более подробно описание проблемы смотрите комментарии в pull request.

В pull request вы так же найдете комментарии с рекомендациями как **можно улучшить** код. Их делать не обязательно, однако в будущем они могут помочь избежать ошибок и сделать код лучше.

**Напомню, чтобы работа была принята нужно исправить все замечания помеченные как **необходимо исправить**.

Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка F12) перед отправкой на ревью. 

Удачных правок!

Ekaterina-ux committed:
Исправлено:
- Обновление токена в запросах с использованием accessToken
Улучшено:
- accessToken, refreshToken в константах
- кастомный хук useForm для контроля инпутов в формах
- Декомпозиция профиля: навигация и форма

andryxxa93 reviewed (approved):
Супер, вы отлично поработали!
Поздравляю! Работа принята.
Желаю удачного продолжения обучения!
